sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala

```diff
@@ -17,391 +17,20 @@
-    val maxQueueSize = 65536
-
-    def apply[T](nonzeroException: Boolean): BlockingStreamed[T] = {
-      val q = new LinkedBlockingQueue[Either[Int, T]](maxQueueSize)
-
-      def next(): Stream[T] = q.take match {
-        case Left(0) => Stream.empty
-        case Left(code) =>
-          if (nonzeroException) scala.sys.error("Nonzero exit code: " + code) else Stream.empty
-        case Right(s) => Stream.cons(s, next())
-      }
-
-      new BlockingStreamed((s: T) => q put Right(s), code => q put Left(code), () => next())
-    }
-  }
-
-  // See scala.sys.process.ProcessImpl.Spawn
-  private object Spawn {
-    def apply(f: => Unit): Thread = apply(f, daemon = false)
-    def apply(f: => Unit, daemon: Boolean): Thread = {
-      val thread = new Thread() { override def run() = { f } }
-      thread.setDaemon(daemon)
-      thread.start()
-      thread
-    }
-  }
-
-  def apply(command: Seq[String]): Stream[String] = {
-    val streamed = BlockingStreamed[String](true)
-    val process = command.run(BasicIO(false, streamed.process, None))
-    Spawn(streamed.done(process.exitValue()))
-    streamed.stream()
-  }
-}
-
-class Dsdgen(dsdgenDir: String) extends Serializable {
-  private val dsdgen = s"$dsdgenDir/dsdgen"
-
-  def generate(
-      sparkContext: SparkContext,
-      tableName: String,
-      partitions: Int,
-      scaleFactor: Int): RDD[String] = {
-    val generatedData = {
-      sparkContext.parallelize(1 to partitions, partitions).flatMap { i =>
-        val localToolsDir = if (new java.io.File(dsdgen).exists) {
-          dsdgenDir
-        } else if (new java.io.File(s"/$dsdgen").exists) {
-          s"/$dsdgenDir"
-        } else {
-          throw new IllegalStateException(
-            s"Could not find dsdgen at $dsdgen or /$dsdgen. Run install")
-        }
-
-        // NOTE: RNGSEED is the RNG seed used by the data generator. Right now, it is fixed to 100.
-        val parallel = if (partitions > 1) s"-parallel $partitions -child $i" else ""
-        val commands = Seq(
-          "bash", "-c",
-          s"cd $localToolsDir && ./dsdgen -table $tableName -filter Y -scale $scaleFactor " +
-          s"-RNGSEED 100 $parallel")
-        BlockingLineStream(commands)
-      }
-    }
-
-    generatedData.setName(s"$tableName, sf=$scaleFactor, strings")
-    generatedData
-  }
-}
+import org.apache.spark.sql.types.StructType
 
 class TPCDSTables(sqlContext: SQLContext, dsdgenDir: String, scaleFactor: Int)
-  extends TPCDSSchema with Logging with Serializable {
+  extends TableGenerator with TPCDSSchema with Logging with Serializable {
 
-  private val dataGenerator = new Dsdgen(dsdgenDir)
-
-  private def tables: Seq[Table] = tableColumns.map { case (tableName, schemaString) =>
+  override protected val dataGenerator: DataGenerator = new Dsdgen(dsdgenDir)
+  override protected val sparkSQLContext: SQLContext = sqlContext
+  override protected val tpcScaleFactor: Int = scaleFactor
```

Comment:
```suggestion
```

Reply:
Please declare `sqlContext: SQLContext` and `scaleFactor: Int` in `TableGenerator`.

Reply:
Updated. Thanks.